clippy_lints/src/dereference.rs

```diff
@@ -457,139 +538,363 @@ fn deref_method_same_type(result_ty: Ty<'_>, arg_ty: Ty<'_>) -> bool {
+        };
+    }
+}
+
+// Checks whether a type is inferred at some point.
+// e.g. `_`, `Box<_>`, `[_]`
+fn ty_contains_infer(ty: &hir::Ty<'_>) -> bool {
+    match &ty.kind {
+        TyKind::Slice(ty) | TyKind::Array(ty, _) => ty_contains_infer(ty),
+        TyKind::Ptr(ty) | TyKind::Rptr(_, ty) => ty_contains_infer(ty.ty),
+        TyKind::Tup(tys) => tys.iter().any(ty_contains_infer),
+        TyKind::BareFn(ty) => {
+            if ty.decl.inputs.iter().any(ty_contains_infer) {
+                return true;
+            }
+            if let FnRetTy::Return(ty) = &ty.decl.output {
+                ty_contains_infer(ty)
+            } else {
+                false
+            }
+        },
+        &TyKind::Path(
+            QPath::TypeRelative(_, path)
+            | QPath::Resolved(
+                _,
+                Path {
+                    segments: [.., path], ..
+                },
+            ),
+        ) => path.args.map_or(false, |args| {
+            args.args.iter().any(|arg| match arg {
+                GenericArg::Infer(_) => true,
+                GenericArg::Type(ty) => ty_contains_infer(ty),
+                _ => false,
+            })
+        }),
+        TyKind::Path(_) | TyKind::OpaqueDef(..) | TyKind::Infer | TyKind::Typeof(_) | TyKind::Err => true,
+        TyKind::Never | TyKind::TraitObject(..) => false,
+    }
+}
+
+// Checks whether a type is stable when switching to auto dereferencing,
+fn param_auto_deref_stability(ty: Ty<'_>) -> Position {
+    let ty::Ref(_, mut ty, _) = ty.kind() else {
+        return Position::Other;
+    };
+
+    loop {
+        break match ty.kind() {
+            ty::Ref(_, ref_ty, _) => {
+                ty = ref_ty;
+                continue;
+            },
+            ty::Bool
+            | ty::Char
+            | ty::Int(_)
+            | ty::Uint(_)
+            | ty::Float(_)
+            | ty::Foreign(_)
+            | ty::Str
+            | ty::Array(..)
+            | ty::Slice(..)
+            | ty::RawPtr(..)
+            | ty::FnDef(..)
+            | ty::FnPtr(_)
+            | ty::Closure(..)
+            | ty::Generator(..)
+            | ty::GeneratorWitness(..)
+            | ty::Never
+            | ty::Tuple(_)
+            | ty::Projection(_) => Position::DerefStable,
+            ty::Infer(_)
+            | ty::Error(_)
+            | ty::Param(_)
+            | ty::Bound(..)
+            | ty::Opaque(..)
+            | ty::Placeholder(_)
+            | ty::Dynamic(..) => Position::ReborrowStable,
+            ty::Adt(..) if ty.has_placeholders() || ty.has_param_types_or_consts() => Position::ReborrowStable,
+            ty::Adt(..) => Position::DerefStable,
+        };
+    }
+}
+
+fn ty_contains_field(ty: Ty<'_>, name: Symbol) -> bool {
+    if let ty::Adt(adt, _) = *ty.kind() {
+        adt.is_struct() && adt.non_enum_variant().fields.iter().any(|f| f.name == name)
```

Comment:
When first checking if it is a struct, you can also just call [`all_fields`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/ty/adt/struct.AdtDef.html#method.all_fields) here, which gives you an iterator over the fields directly.
```suggestion
        adt.is_struct() && adt.all_fields().any(|f| f.name == name)
```